providers/azure.go

```diff
@@ -233,13 +234,50 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 	return email, err
 }
 
-// GetEmailAddress returns the Account email address
-func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
+// full claims that can be emitted by AAD are
+// https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens
+// only include email here since we only need email extraction
+type aadIDTokenClaims struct {
+	Email string `json:"email,omitempty"`
+}
+
+// Valid is the function validating the claims,
+// required by jwt.Claims interface.
+// returning nil as we only use it to extract claims instead of validating them
+func (aadIDTokenClaims) Valid() error {
+	return nil
+}
+
+func getEmailFromIDToken(idToken string) (string, error) {
+	claims := &aadIDTokenClaims{}
+	parser := jwt.Parser{}
+
+	// ParseUnverified to extract claims without verifying signature
+	if _, _, err := parser.ParseUnverified(idToken, claims); err != nil {
```

Comment:
Are you able to get this to an `oidc.IDToken` via an `oidc.Verifier` ([sample](https://github.com/oauth2-proxy/oauth2-proxy/blob/3ff0c23a9e418875cdd7460b0a6a5b40dea71770/providers/oidc_test.go#L151)) and use the `Claims` method like here: https://github.com/oauth2-proxy/oauth2-proxy/blob/3ff0c23a9e418875cdd7460b0a6a5b40dea71770/providers/oidc.go#L233

I figure I'll eventually be on the hook for generic claim -> header support across all providers. When that happens, I'll need to consolidate logic. This would be the 3rd different way to extract IDToken claims, so I'd like to consolidate, at least onto the same `oidc` library if we aren't refactoring the logic into a dedicated method in `utils` at this point.

Reply:
both `oidc` and `jwt` libraries are doing token validations, and I'm with you, totally agree that these need to be consolidated. However, what's not clear to me is the responsibility of each provider, and when this validation should be consolidated into. The code I refactored here is the one without any dependency such as jwks.
how do you think? perhaps leave the consolidation out of this PR?

Reply:
I'm leaning on you a little bit since you have a good understanding of Azure & the OIDC specifics of it 😄 

When I take on a lot of these refactors in the future for new features, Azure isn't a provider I use myself. So I won't have a test account or know the nuances of it.

That's why I'm looking to get the guts of what we need to consolidate on as close as possible & have strong test cases that emulate Azure IDTokens well so in a future refactor I can spruce up code and rely on existing tests passing to be confident in an changes to the Azure provider.

Reply:
At the very least, just use the `oidc.Verifier` I think -- you would need the same guts as the struct & `Valid` method you made for the `jwt` library right?

At least that way, the foundation is in place with the right `issuer` & `audience` values in code.

Reply:
@JoelSpeed -- @weinong and I discussed offline: I don't think we should take any data from unverified tokens to be used in our auth critical `Email` or `Groups` session fields.

@weinong will wait until we wrap up this PR: https://github.com/oauth2-proxy/oauth2-proxy/pull/869

With that in place, if an Azure user wants this, they can setup the OIDC config options that would results in an `oidc.IDTokenVerifier` on the ProviderData struct. (as an added benefit this makes Azure bearer token authentication ready where it wasn't before).

If that is non-nil, this logic can proceed and extract claims from the IDToken to get an email (i.e. a user has setup the proper configs for Azure to say OIDC IDToken extraction is a safe route for identity extraction). Otherwise it will skip it and rely on the previous `ProfileURL` based methods to get an Email.